src/backend/downloadmanager/downloadqueue.ts

```diff
@@ -0,0 +1,160 @@
+import { logError, logInfo, LogPrefix } from '../logger/logger'
+import { getMainWindow } from '../utils'
+import Store from 'electron-store'
+import { DMQueueElement } from 'common/types'
+import { installQueueElement } from './utils'
+
+const downloadManager = new Store({
+  cwd: 'store',
+  name: 'download-manager'
+})
+
+/* 
+#### Private ####
+*/
+
+type DownloadManagerState = 'idle' | 'running'
+let queueState: DownloadManagerState = 'idle'
+
+function getFirstQueueElement() {
+  if (downloadManager.has('queue')) {
+    const elements = downloadManager.get('queue') as DMQueueElement[]
+    return elements.at(0)
+  }
+
+  return null
+}
+
+function addToFinished(element: DMQueueElement) {
+  if (!element) {
+    logError('Can not add undefined element to finished!', {
+      prefix: LogPrefix.DownloadManager
+    })
+    return
+  }
+
+  let elements: DMQueueElement[] = []
+  if (downloadManager.has('finished')) {
+    elements = downloadManager.get('finished') as DMQueueElement[]
+  }
+
+  const elementIndex = elements.findIndex(
+    (el) => el.params.appName === element.params.appName
+  )
+
+  if (elementIndex >= 0) {
+    elements[elementIndex] = element
+  } else {
+    elements.push(element)
+  }
+
+  downloadManager.set('finished', elements)
+  logInfo([element.params.appName, 'added to download manager finished.'], {
+    prefix: LogPrefix.DownloadManager
+  })
+}
+
+/* 
+#### Public ####
+*/
+
+async function initQueue() {
+  const window = getMainWindow()
+  let element = getFirstQueueElement()
+  queueState = element ? 'running' : 'idle'
+
+  while (element) {
+    // need to type cohert here since this method can return string as well
+    const { status } = (await installQueueElement(window, element.params)) as {
```

Comment:
do we need to pass the windows here? I don't know how this will work with the `close heroic to the tray` feature (I think it destroys the main window and creates a new one when re-opened, so this will be pointing to the wrong place)

I think we should get the main window inside the installQueueElement method when it's needed instead

Reply:
I read a bit more about this and looks like `close to try` only hides the window so the reference is not broken, so this shouldn't be an issue.